### PR TITLE
simple functorch with modes

### PR DESCRIPTION
**Pull request by samdow**
This adds a rewrite of the simple functorch notebook to use torch dispatch modes. In doing this, we needed to add padding to vmap (i.e. we need to make `vmap(f)(torch.randn(2), torch.randn(2, 4))` work)

Still TODO:
- [ ] fix bugs identified in #31 

It raised a couple bugs in torch dispatch modes
- [ ] fix bug to set ancestors if inner is explicitly given
- [ ] fix bug to make `enable_torch_dispatch_mode(None, replace=True)` work

### CODE REVIEW

File: simple_functorch_modes.py [hunk 1]
```diff
@@ -0,0 +1,1302 @@
...
+assert(isinstance(z, torch.Tensor))
+
+"""With the Dispatcher structure in hand, we are now in a good place to
+port the autograd implementation from Simple Autograd into our new
+framework.
+"""
+
+from torch.utils._mode_utils import no_dispatch
+
+class Autograd(Dispatcher):
```
Comment (samdow) [inline simple_functorch_modes.py:362]: Note that because we're sitting at the torch dispatch level, autograd has already happened so we still have to rewrite autograd here and can't reuse PyTorch's autograd. This would be a no-go in actual functorch so we'll need to figure out a solution here
